### ISSUE TITLE
BR support table route when restore 

### ISSUE THREAD

**Issue by 3pointer** [pingcap/br#252]
### Describe your feature request related problem:

when user already has a table A in cluster.
sometimes if table A partially broken.
user need recover table A from BR but don't want to drop origin table A. 
so a kind of `table route` function is needed.

### Describe the feature you'd like:

table route in restore.

### Describe alternatives you've considered:

not sure if we need introduce config file to support this funciton.

**Comment by kennytm** [comment]
if we don't need wildcard support we could reuse the syntax of [--replicate-rewrite-db](https://dev.mysql.com/doc/refman/8.0/en/replication-options-slave.html#option_mysqld_replicate-rewrite-db).

```sh
./br restore full --rename-table 'db1.tbl1->db2.tbl2' ...
```

Given the scenario, perhaps even just `--rename-db 'db1->db2'` is sufficient.

It is probably unsafe for the renamed target tables to have name collision. That is, we shouldn't support shard merging in BR (`--rename-table 'a.t1->a.t,a.t2->a.t'`).

Note that the rename operation needs to be done for the incremental DDLs.

**Comment by YuJuncen** [comment]
Comparing to mysql `--replicate-rewrite-db` syntax, some of us prefer docker-style syntax, and finally we decided to use it. Which is:

1. use `--rename-table "originalDB.tbl:rewritedDB.tbl2"` or `--rename-db "originalDB:rewritedDB"` to make a route rule.
2. if there are multi route rules, use multi `--route` flags to make route rules.
3. if there are special text in database or table name, use backquote to enclose them, like: `` `db.1`.`table-name` ``.

The implementation of this function can be split into several steps:

1. make a parser that can parse this flag into some data structures, at this time, we can also check whether there is any name collision.
2. then, we can rewrite db/table names after we get `backupmeta` from storage, we need rewrite two space at `backupmeta`:
  a. tables at `backupMeta.Schemas`.
  b. DDLs jobs at `backupMeta.Ddls`(for incremental restore).

in the step (b) those fields of `rc.ddlJobs[*]` should be rewritten:
  - `TableInfo` and `DBInfo` in `BinlogInfo` field.
  - `SchemaName`
  -  the table names in the `Query` string.
  The `Query` string is just a SQL statement. We can use the `Visitor` interface of our SQL parser to rewrite them, a code snippet is [here](https://gist.github.com/YuJuncen/435c69c9cc1ec4ccb8c71ba22d08683c).

Then, let the restore process go on and our route rules should be enabled.

**Comment by kennytm** [comment]
What to do when user specifies both `--rename-table` and `--rename-db` flags?

```
--rename-table 'a.b:c.d' --rename-db 'a:x' --rename-table 'a.m:d.n' --rename-db 'c:y'
```

**Comment by YuJuncen** [comment]
@kennytm We can chain them as a "filter chain", for each table, apply rules sequentially.

e.g., for your example, we can make the execution chain: 
```
(1)a.b:c.d -> 
(2)a:x -> 
(3)a.m:d.n -> 
(4)c:y
```

when accepting `a.b`: 
(1) applies, make it to be `c.d`;
(2) cannot apply to it, because it is rewritten to `c.d`;
(3) cannot apply to it,
(4) applies, make it to be `y.d`

I think the chain method is easy to implement and understand. For implementation, we can write a trait:
```go
interface RouteRule {
    Rewrite(db, table model.CIStr) (rdb, rtable model.CIStr)
}
```

and implement it by atom rules `DBRouteRule`, `TableRouteRule`, `NilRouteRule` and combinator `AndRouteRule`:
```go
struct AndRouteRule {
    car RouteRule 
    cdr RouteRule
}

func (arr AndRouteRule) Rewrite(db, table model.CIStr) (rdb, rtable model.CIStr) {
    rdb, rtable = arr.car.Rewrite(db, table)
    return arr.cdr.Rewrite(rdb, rtable)
}
```

**Comment by kennytm** [comment]
@YuJuncen for the chain to work, `pflags` need to be modified to recognize `--rewrite-table` and `--rewrite-db` at the same level

**Comment by YuJuncen** [comment]
@kennytm Yes... That is a problem, how about just merge them into a single `--rewrite` flag?

**Comment by kennytm** [comment]
Which means

```
--rename 'a.b:c.d' --rename 'a:x' --rename 'a.m:d.n' --rename 'c:y'
```

**Comment by YuJuncen** [comment]
> Which means
> 
> ```
> --rename 'a.b:c.d' --rename 'a:x' --rename 'a.m:d.n' --rename 'c:y'
> ```

yep
